File: Kakao/2019_블록 게임.cpp
// 출제의도: "구현"

#include <string>
#include <vector>
#include <iostream>

using namespace std;

typedef pair<int,int>   Pos;

struct Type {
    vector<Pos> body;
    vector<Pos> check;
};

struct Block {
    int type;
    Pos top;
    Block(int Type, Pos Top):type(Type),top(Top){}
};

vector<Type> type(5);

bool isOutOfBound(vector<vector<int>> &board, Pos pos) {
    int N = board.size();
    if(pos.first<0 || pos.first>=N || pos.second<0 || pos.second>=N)   return true;
    return false;
}

int getType(vector<vector<int>> &board, Pos pos) {
    int no = board[pos.first][pos.second];
    for(int i=0;i<type.size();i++) {
        int cnt=1;
        for(int j=0;j<type[i].body.size();j++) {
            Pos next = Pos(pos.first+type[i].body[j].first,pos.second+type[i].body[j].second);
            if(!isOutOfBound(board,next) && board[next.first][next.second]==no)  cnt++;
        }
        if(cnt==4)  return i;
    }
    return -1;
}

bool isPossibleToPop(vector<vector<int>> &board, Block block) {
    if(block.type==-1)  return false;
    
    for(int i=0;i<type[block.type].check.size();i++) {
        Pos check = Pos(block.top.first+type[block.type].check[i].first,block.top.second+type[block.type].check[i].second);
        for(int i=check.first;i>=0;i--) {
            if(board[i][check.second]!=0) {
                return false;
            }
        }
    }
    return true;
}

void pop(vector<vector<int>> &board, Block block) {
    board[block.top.first][block.top.second] = 0;
    for(int i=0;i<type[block.type].body.size();i++) { 
        board[block.top.first+type[block.type].body[i].first][block.top.second+type[block.type].body[i].second] = 0;
    }
}

int solution(vector<vector<int>> board) {
    int answer = 0;
    int answer_prev = -1;
    vector<Block> blocks;
    
    type[0].body = {Pos(1,0),Pos(1,1),Pos(1,2)};
    type[0].check = {Pos(0,1),Pos(0,2)};
    type[1].body = {Pos(1,0),Pos(2,0),Pos(2,-1)};
    type[1].check = {Pos(0,-1),Pos(1,-1)};
    type[2].body = {Pos(1,0),Pos(2,0),Pos(2,1)};
    type[2].check = {Pos(0,1),Pos(1,1)};
    type[3].body = {Pos(1,0),Pos(1,-1),Pos(1,-2)};
    type[3].check = {Pos(0,-1),Pos(0,-2)};
    type[4].body = {Pos(1,0),Pos(1,-1),Pos(1,1)};
    type[4].check = {Pos(0,1),Pos(0,-1)};
    
    for(int i=0;i<board.size();i++) {
        for(int j=0;j<board[0].size();j++) {
            int type = getType(board,Pos(i,j));
            if(board[i][j]!=0 && type!=-1) {
                blocks.push_back(Block(type,Pos(i,j)));
            }
        }
    }
    while(answer!=answer_prev)
    {
        answer_prev = answer;
        for(int i=0;i<blocks.size();i++) {
            if(isPossibleToPop(board,blocks[i])) {
                answer++;
                pop(board,blocks[i]);
                blocks[i].type=-1;
            }
        }
    }
    return answer;
}

File: Kakao/2019_블록 게임.swift
// 출제의도: "구현"
// 1) swift에서 struct는 생성된 이후 property 수정이 불가. 따라서, mutating func을 정의하여 수정해야 함.
// 2) for a in b 문에서 a는 let으로 받게 됨. 따라서, 수정이 불가능. 이 문제를 피하기 위해 for i in 0..<b.count로 인덱스로 접근해야 함.
// 3) 지울 수 있는 블록의 패턴은 5가지. lefttop을 기준으로 모양을 결정할 수 있음.
// 4) 이 문제는 Swift가 C++보다 풀기에 적합하지 않음. 


import Foundation

struct Type {
    let no: Int
    let body: [(Int, Int)]
    let dent: [(Int, Int)]
    
    init(_ no: Int, _ body: [(Int, Int)], _ dent: [(Int, Int)]) {
        self.no = no
        self.body = body
        self.dent = dent
    }
}

struct Block {
    let leftTop: (Int, Int)
    let type: Int
    var isPoped = false
    
    init(_ leftTop: (Int, Int), _ type: Int) {
        self.leftTop = leftTop
        self.type = type
    }
    
    mutating func pop()->Void {
        self.isPoped = true
    }
}

let types: [Type] = [Type(0,[(1,0),(1,1),(1,2)],[(0,1),(0,2)]),
                     Type(1,[(1,0),(2,0),(2,-1)],[(0,-1),(1,-1)]),
                     Type(2,[(1,0),(2,0),(2,1)],[(0,1),(1,1)]),
                     Type(3,[(1,0),(1,-1),(1,-2)],[(0,-1),(0,-2)]),
                     Type(4,[(1,0),(1,-1),(1,1)],[(0,-1),(0,1)])]

func isOutOfBound(_ N: Int, _ x: Int, _ y: Int) -> Bool {
    if x<0 || x>=N || y<0 || y>=N {
        return true
    }
    return false
}

func popBlock(_ board: inout [[Int]], _ block: Block) -> Void {
    board[block.leftTop.0][block.leftTop.1] = 0;
    types[block.type].body.forEach {
        board[block.leftTop.0 + $0.0][block.leftTop.1 + $0.1] = 0
    }
}

func solution(_ board:[[Int]]) -> Int {
    var board = board
    var answer: Int = 0
    var blocks = [Block]()
    var visited = [Int: Bool]()
    var N = board.count
    
    for i in 0..<N {
        for j in 0..<N {
            if board[i][j] == 0 || visited[board[i][j], default: false] {
                continue
            }
            visited[board[i][j]] = true
            
            for type in types {
                var fnd = true
                type.body.forEach {
                    if isOutOfBound(N,i+$0.0,j+$0.1) || board[i][j] != board[i+$0.0][j+$0.1]{
                        fnd = false
                        return
                    }
                }
                if fnd {
                    blocks.append(Block((i,j),type.no))
                    break
                }
            }
        }
    }
    
    for i in 0..<blocks.count {
        var numOfPop: Int = 0

        for i in 0..<blocks.count {
            if blocks[i].isPoped {
                continue
            }

            let dents = types[blocks[i].type].dent
            var isEmptyVerticaly = true

            dents.forEach {
                for r in 0...blocks[i].leftTop.0+$0.0 {
                    if board[r][blocks[i].leftTop.1+$0.1] > 0 {
                        isEmptyVerticaly = false
                        return
                    }
                }
            }

            if isEmptyVerticaly {
                numOfPop += 1
                popBlock(&board,blocks[i])
                blocks[i].pop()
            }
        }

        if numOfPop == 0 {
            break
        }
        answer += numOfPop
    }
    
    return answer
}
